clang/lib/Parse/ParseTemplate.cpp

```diff
@@ -1779,6 +1779,42 @@ void Parser::checkPotentialAngleBracket(ExprResult &PotentialTemplateName) {
                     Priority);
 }
 
+bool Parser::isMissingTemplateKeywordBeforeScope(bool AnnotateInvalid) {
+  assert(Tok.is(tok::coloncolon));
+  Sema::DisableTypoCorrectionRAII DTC(Actions);
+  ColonProtectionRAIIObject ColonProtection(*this);
+
+  SourceLocation StartLoc = Tok.getLocation();
+  if (TryAnnotateTypeOrScopeToken())
+    return true;
+  if (Tok.isSimpleTypeSpecifier(getLangOpts()))
+    return false;
+  CXXScopeSpec SS;
+  ParseOptionalCXXScopeSpecifier(SS, /*ObjectType=*/nullptr,
+                                 /*ObjectHasErrors=*/false,
+                                 /*EnteringContext=*/false);
+  ExprResult Result = tryParseCXXIdExpression(SS, /*isAddressOfOperand=*/false);
```

Comment:
I'm worried that this can perform semantic actions (for example, template instantiations) that we're not permitted to perform, resulting in errors outside the immediate context, instantiation of variables with initializers that would incorrectly run on program startup, or other misbehavior. Generally-speaking, it's not conforming to try to parse and semantically check code and then backtrack if it didn't work. We can do such things on error recovery paths, and we can do purely syntactic tentative parsing things (where it's OK to ask Sema about name lookup, but not OK to ask it to do more complicated things), but we can't do full semantic analysis tentatively.

Reply:
For a well-formed program, any tentative parsing done here will happen anyways when parsing the second operand of the `>` operator. Only for an _ill-formed_ program (i.e. one where `template` is missing) do we instantiate things that should not be instantiated.

As an alternative, we could also just lookup the name immediately following the `::` _nested-name-specifier_ and make the determination based on that.

Reply:
Parsing the same expression twice could also be problematic in some cases -- for example, if it contains a lambda, we might assign it the wrong mangling number. Can we arrange things such that for a valid program, we only parse the expression once?

Reply:
> Parsing the same expression twice could also be problematic in some cases -- for example, if it contains a lambda, we might assign it the wrong mangling number. Can we arrange things such that for a valid program, we only parse the expression once?

@zygoloid That already is the case:
```cpp
template<int I>
struct A
{
    constexpr static int x = 0;

    template<typename T>
    using B = A;
};

template<typename T>
struct B
{
    constexpr static int x = 0;
};

template<typename T>
void f(T t)
{
    t.A<0>::template B<decltype([]() { })>::x; // ok, interpreted as '((t.A) < 0) > ::template B<decltype([]() { })>::x'
                                      // lambda expression is only parsed once

    t.A<0>::template C<decltype([]() { })>::x; // warning: use 'template' keyword to treat 'A' as a dependent template name
                                               // lambda expression is parsed twice
}
```

Reply:
Thanks for helping me understand :-) I misunderstood your previous comment (any tentative parsing done here will happen anyways when parsing the second operand of the > operator) as suggesting we'd sometimes parse and type-check the same thing twice in a non-ill-formed program. It sounds like we're not doing that, which satisfies my concern here.

Reply:
@zygoloid Do you otherwise think that it's reasonable to apply this change to _all_ C++ language modes (given the existence of [this thread](https://lists.isocpp.org/core/2024/07/16028.php) on the core reflector resulting from the initial application of this patch _without_ the enhanced identification of missing `template`)?